platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/AjaxEngine.ts

```diff
@@ -0,0 +1,166 @@
+	// initializes the body object of this document
+	public initializeBody(nodeId: number) {
+		this.initializeNode(nodeId, document.body);
+	}
+
+	// executes the given Javascript code
+	public executeScriptCode(scriptCode: string) {
+		eval(scriptCode);
+	}
+
+	// creates a new HTML element
+	public createElement(nodeId: number, tag: string) {
+		let node = document.createElement(tag);
+		this.initializeNode(nodeId, node);
+
+		// if this is a value node, approve its current value
+		switch(tag.toUpperCase()) {
+		case 'INPUT':
+		case 'TEXTAREA':
+		case 'SELECT':
+			VALUE_NODE_MAP.addNode(node);
+		}
+	}
+
+	// creates a new text node
+	public createTextNode(nodeId: number, text: string) {
+		let node = document.createTextNode(text);
+		this.nodes.set(nodeId, node);
+	}
+
+	// appends the specified child to the parent
+	public appendChild(parentId: number, childId: number) {
+		this.getElement(parentId).appendChild(this.getNode(childId));
+	}
+
+	// appends the specified child to the parent
+	public insertBefore(parentId: number, childId: number, otherChildId: number) {
+		this.getElement(parentId).insertBefore(this.getNode(childId), this.getNode(otherChildId));
+	}
+
+	// removes the specified child from the parent
+	public removeChild(parentId: number, childId: number) {
+		this.getElement(parentId).removeChild(this.getNode(childId));
+	}
+
+	// replaces the old child of the parent with the new child
+	public replaceChild(parentId: number, newChildId: number, oldChildId: number) {
+		this.getElement(parentId).replaceChild(this.getNode(newChildId), this.getNode(oldChildId));
+	}
+
+	public setAttribute(nodeId: number, attribute: string, value: string) {
+		this.getElement(nodeId).setAttribute(attribute, value);
+	}
+	
+	public listenToEvent(nodeID: number, event: string, doListen: boolean) {
+		let element = this.getElement(nodeID);
+		if(element == null)
+			return;
+	
+		let handler = doListen? handleDomEvent : null;
+	
+		switch(event)
+		{
+		case 'CLICK':       element.onclick = handler; break;
+		case 'CHANGE':      CHANGE_EVENT_MANAGER.setListenToChangeEvent(element, doListen); break;
+		case 'CONTEXTMENU': element.oncontextmenu = doListen? (event => {handleDomEvent(event); event.preventDefault();}) : null; break;
+		case 'DBLCLICK':    element.ondblclick = handler; break;
+		case 'ENTER':       KEYBOARD_EVENT_MANAGER.setListenToKey(element, event, doListen); break;
+		case 'ESCAPE':      KEYBOARD_EVENT_MANAGER.setListenToKey(element, event, doListen); break;
+		case 'KEYPRESS':    element.onkeypress = handler; break;
+		case 'SPACE':       KEYBOARD_EVENT_MANAGER.setListenToKey(element, event, doListen); break;
+		case 'TAB':         KEYBOARD_EVENT_MANAGER.setListenToKey(element, event, doListen); break;
+		default: alert('Unknown event ' + event + '.');
+		}
+	}
+
+	private initializeNode(nodeId: number, node: HTMLElement) {
+		node.id = 'n' + nodeId;
+		this.nodes.set(nodeId, node);
+	}
+}
+
+const _DOM_CONTEXT_ = new AjaxEngine();
```

Comment:
...if it's in fact the engine, we should probably just rename the constant to `_ENGINE_` or `_AJAX_ENGINE_`. Because what's the point of having a synonym?

Reply:
the point is that I am waiting for your PR to get ready :P `_DOM_CONTEXT_` is used in `DomPopupEngine.js`

Reply:
i see